File: src/typescript/bashkit-driver.ts
/**
 * BashkitDriver: resolves to BashkitCLIDriver.
 */

import { execSync } from "child_process";
import {
  type ShellDriver,
  ShellDriverFactory,
  type ShellDriverOptions,
} from "./drivers.js";
import {
  BashkitCLIDriver,
  type BashkitCLIDriverOptions,
} from "./bashkit-cli-driver.js";

export class BashkitDriver {
  /**
   * Return a bashkit ShellDriver backed by the CLI subprocess.
   */
  static resolve(opts: BashkitCLIDriverOptions = {}): ShellDriver {
    // If test override provided, use it directly
    if (opts._execOverride) {
      return new BashkitCLIDriver(opts);
    }
    // Check if bashkit binary is available
    if (!BashkitDriver._checkCli()) {
      throw new Error(
        "bashkit not found — install with: cargo install bashkit-cli"
      );
    }
    return new BashkitCLIDriver(opts);
  }

  /** Check if bashkit is on PATH. */
  private static _checkCli(): boolean {
    try {
      execSync("which bashkit", { stdio: "ignore" });
      return true;
    } catch {
      return false;
    }
  }
}

/**
 * Register the "bashkit" driver with ShellDriverFactory.
 */
export function registerBashkitDriver(): void {
  ShellDriverFactory.register("bashkit", (opts?: ShellDriverOptions) =>
    BashkitDriver.resolve(opts as BashkitCLIDriverOptions)
  );
}

File: src/python/bashkit_driver.py
"""BashkitDriver: resolves to BashkitPythonDriver when bashkit package is available."""

from __future__ import annotations

from typing import Any

from src.python.drivers import ShellDriver, ShellDriverFactory


class BashkitDriver:
    """Resolves the bashkit driver (requires bashkit Python package)."""

    @staticmethod
    def resolve(**kwargs: Any) -> ShellDriver:
        """Return a BashkitPythonDriver if bashkit is installed."""
        # Allow test override to skip import check
        bash_override = kwargs.get("bash_override")
        if bash_override is not None:
            from src.python.bashkit_python_driver import BashkitPythonDriver
            return BashkitPythonDriver(**kwargs)
        try:
            import bashkit  # noqa: F401
        except ImportError:
            raise RuntimeError(
                "bashkit not found — install with: pip install bashkit"
            )
        from src.python.bashkit_python_driver import BashkitPythonDriver
        return BashkitPythonDriver(**kwargs)


def register_bashkit_driver() -> None:
    """Register the 'bashkit' driver with the ShellDriverFactory."""
    ShellDriverFactory.register("bashkit", lambda **kw: BashkitDriver.resolve(**kw))
